internal/statemachines/activity.go

```diff
@@ -0,0 +1,152 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package statemachines
+
+import (
+	commandpb "go.temporal.io/api/command/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+
+	"go.temporal.io/sdk/internal"
+)
+
+type ActivityStateMachine struct {
+	definition               *StateMachineDefinition
+	activityID               string
+	activityType             internal.ActivityType
+	activityCancellationType ActivityCancelationType
+	activityParams           ExecuteActivityParameters
+	// TODO: Factor out common fields
+	commandSink CommandSink
+}
+
+type ActivityState int16
+
+const (
+	stateCreated ActivityState = iota
+	stateScheduleCommandCreated
+	stateScheduledEventRecorded
+	stateStarted
+	stateCompleted
+	stateFailed
+	stateTimedOut
+	stateCanceled
+	stateScheduledActivityCancelCommandCreated
+	stateScheduledActivityCancelEventRecorded
+	stateStartedActivityCancelCommandCreated
+	stateStartedActivityCancelEventRecorded
+)
+
+func (d ActivityState) Name() string {
+	return d.String()
+}
+
+func (d ActivityState) String() string {
+	switch d {
+	case stateCreated:
+		return "CREATED"
+	case stateScheduleCommandCreated:
+		return "SCHEDULE_COMMAND_CREATED"
+	case stateScheduledEventRecorded:
+		return "SCHEDULED_EVENT_RECORDED"
+	case stateStarted:
+		return "STARTED"
+	case stateCompleted:
+		return "COMPLETED"
+	case stateFailed:
+		return "FAILED"
+	case stateTimedOut:
+		return "TIMED_OUT"
+	case stateCanceled:
+		return "CANCELED"
+	case stateScheduledActivityCancelCommandCreated:
+		return "SCHEDULED_ACTIVITY_CANCEL_COMMAND_CREATED"
+	case stateScheduledActivityCancelEventRecorded:
+		return "SCHEDULED_ACTIVITY_CANCEL_EVENT_RECORDED"
+	case stateStartedActivityCancelCommandCreated:
+		return "STARTED_ACTIVITY_CANCEL_COMMAND_CREATED"
+	case stateStartedActivityCancelEventRecorded:
+		return "STARTED_ACTIVITY_CANCEL_EVENT_RECORDED"
+	default:
+		panic("Attempted to stringify unknown activity state. This should be impossible.")
+	}
+}
+
+// TODO: Replace XX. prefixes with different types / constants etc
+func NewActivityStateMachine() *ActivityStateMachine {
+	fsmDef := BuildStateMachine("ActivityStateMachine", stateCreated, []State{stateCompleted, stateFailed, stateTimedOut, stateCanceled})
+	fsmDef.add(stateCreated, ExplicitEventSchedule, stateScheduleCommandCreated)
+
+	fsmDef.add(stateScheduleCommandCreated, CommandEvent{enumspb.COMMAND_TYPE_SCHEDULE_ACTIVITY_TASK}, stateScheduleCommandCreated)
```

Comment:
Why not have addCommand and addEvent methods instead?

Reply:
That totally works for me. I was trying to stick to the Java semantics, but the fact that I have to use these wrapper structs since there are no generics kinda defeats the point.

In fact, may as well have `addExplicitEvent` be separate too.